Fix #102: Move exclusive vm access logic into OMR

youngar opened the pull request:
There are 4 new callbacks for `EnvironmentLanguageInterface` related to 'critical' native code.  When a thread has critical heap access, heap objects are not allowed to move.  This means, for example that a generational scavenge or heap compaction can not occur.
```C++
// Release critical access if you have it.
MM_EnvironmentLanguageInterface::releaseCriticalHeapAccess();
// Reaquire critical access if you had it before calling reacquireCriticalHeapAccess.
MM_EnvironmentLanguageInterface::reacquireCriticalHeapAccess();

exclusiveAccessForGCBeatenByOtherThread();
exclusiveAccessForGCObtainedAfterBeatenByOtherThread();
```
I'm open for any ideas to improve this, or rename these functions.  We could also defer fixing these to a future item for improving our critical heap access support.


There a call to `GC_OMRVMInterface::flushCachesForGC()` in `acquireExclusiveVMAccessForGC()` and `tryAcquireExclusiveVMAccessForGC()` which may or may not be bogus.

charliegracie commented on gc/base/EnvironmentLanguageInterface.hpp:
```diff
@@ -53,156 +52,89 @@ class MM_EnvironmentLanguageInterface : public MM_BaseVirtual
+	virtual void
+	releaseExclusiveVMAccess()
+	{
+		this->releaseVMAccess();
```
When a thread releases exclusive VM access it needs to release VM Access?

charliegracie commented on gc/base/EnvironmentLanguageInterface.hpp:
```diff
@@ -53,31 +52,38 @@ class MM_EnvironmentLanguageInterface : public MM_BaseVirtual
+	virtual void
+	acquireExclusiveVMAccess()
+	{
+		this->acquireVMAccess();
```
Is this the correct call?

youngar commented:
The threads should not have been modifying their shared VM access while acquiring or releasing exclusive VM access.  I have fixed this.

I have also updated the implementation of MM_EnvironmentLanguageInterface::acquireExclusiveVMAccess() to grab the _vmThreadListMutex, and release it in MM_EnvironmentLanguageInterface::releaseExclusiveVMAccess().

amicic commented:
In MM_EnvironmentLanguageInterface::acquireExclusiveVMAccess() should we increment omrThread->exclusiveCount after acquiring vmThreadListMutex?

amicic commented:
Not directly related to this change, but i'm unsure why we have exclusiveCount in both OMR Thread and MM_EnvironmentLanguageInterface.

youngar commented:
@amicic 
My impression is that the exclusiveCount in EnvironmentLanguageInterface is "GC-side," used during `EnvironmentLanguageInterface::AcquireExclusiveVMAccessForGC()`, while the exclusiveCount in the OMR_VMThread is "language-side" and meant to be used internally by the language.  In Java, `OMR_VMThread->exclusiveCount` is actively used to manage all exclusive Java VM access.  In all our other OMR languages, it is not really used for any practical purpose.  We may want to eventually move the exclusiveCount from OMR_VMThread, into J9VMThread.

In OMR we commonly assert that we have exclusive VM Access. This is normally done though a call to `Assert_MM_mustHaveExclusiveVMAccess`, which relies on `OMR_VMThread->exclusiveCount` to be properly updated by the language. (`P1` is an OMR_VMThread)
```C
// Assert that the thread has exclusive VM access
TraceAssert=Assert_MM_mustHaveExclusiveVMAccess noEnv Overhead=1 Level=1 Assert="(P1)->exclusiveCount != 0"
```
If we do eventually remove the field `OMR_VMThread->exclusiveCount`, we will have to replace or remove this assertion.

youngar commented:
> In MM_EnvironmentLanguageInterface::acquireExclusiveVMAccess() should we increment omrThread->exclusiveCount after acquiring vmThreadListMutex?

That logically makes sense to me, but in Java and Ruby we increment  `omrThread->exclusiveCount` before acquiring the mutex.  I don't think that there is a functional difference, since `omrThread->exclusiveCount` is only read and written to by the owning thread.

charliegracie commented:
@youngar could the EnvironmentLanguageInterface exclusive count just be removed and everyone only references the OMR_VMThread one?  It does not have be completed as part of this change but I am just curious?

youngar commented:
Right now, the GC will not attempt acquire exclusive VM access recursively through calls to `EnvironmentLanguageInterface::acquireExclusiveVMAccess`. When the GC does recursive calls to `EnvironmentLanguageInterface::acquireExclusiveVMAcessForGC`, threads keep track of their own recursive counts to make sure that exclusive VM access is only obtained and given up once.
 
We could stop using `OMR_VMThread->_exclusiveCount` in Java, and instead use an exclusiveCount off of Java's J9VMThread.  We could then use the `OMR_VMThread->_exclusiveCount` for OMR's GC, and then delete the field in EnvironmentLanguageInterface.

We can't directly use the same exclusive count field in the GC and language.  This is because a call to `EnvironmentLanguageInterface::unwindExclusiveVMAccessForGC` sets the exclusive count to 0 after a GC, and this will also unwind the language's recursive exclusive VM access.

charliegracie commented:
Ok sounds good.
